File: services/gateway/telemetry/heartbeat.go
// Package telemetry provides heartbeat usage reporting for the on-prem gateway.
//
// Collects counter-only metrics (prediction counts, latency percentiles, error counts)
// and periodically reports them to the vendor's cloud control plane.
//
// SECURITY: No sensitive data (ciphertext, features, keys, payloads) is ever
// included in telemetry. Only counters and timing information are reported.
package telemetry

import (
	"encoding/json"
	"fmt"
	"log"
	"math"
	"sort"
	"sync"
	"time"
)

// ForbiddenFields that must NEVER appear in telemetry reports.
var ForbiddenFields = map[string]bool{
	"ciphertext": true, "plaintext": true, "secret_key": true,
	"eval_key": true, "features": true, "predictions": true,
	"payload": true, "api_key": true, "password": true,
	"token": true, "model_weights": true,
}

// PredictionEvent represents a single prediction for telemetry aggregation.
type PredictionEvent struct {
	TenantID  string
	ModelID   string
	LicenseID string
	LatencyMs float64
	Success   bool
	Timestamp time.Time
}

// HeartbeatReport is the aggregated telemetry payload sent to the control plane.
type HeartbeatReport struct {
	TenantID              string         `json:"tenant_id"`
	LicenseID             string         `json:"license_id"`
	ReportID              string         `json:"report_id"`
	IntervalStart         time.Time      `json:"interval_start"`
	IntervalEnd           time.Time      `json:"interval_end"`
	TotalPredictions      int64          `json:"total_predictions"`
	SuccessfulPredictions int64          `json:"successful_predictions"`
	FailedPredictions     int64          `json:"failed_predictions"`
	PredictionsByModel    map[string]int64 `json:"predictions_by_model"`
	LatencyP50Ms          float64        `json:"latency_p50_ms"`
	LatencyP95Ms          float64        `json:"latency_p95_ms"`
	LatencyP99Ms          float64        `json:"latency_p99_ms"`
}

// ToJSON serializes the report, ensuring no forbidden fields are included.
func (r *HeartbeatReport) ToJSON() ([]byte, error) {
	return json.Marshal(r)
}

// ReportCallback is invoked when a new heartbeat report is ready.
type ReportCallback func(report *HeartbeatReport)

// Collector aggregates prediction events and produces periodic reports.
type Collector struct {
	mu sync.Mutex

	tenantID  string
	licenseID string
	interval  time.Duration
	callback  ReportCallback

	events        []PredictionEvent
	intervalStart time.Time
	reportCounter int64

	running bool
	stopCh  chan struct{}
	reports []*HeartbeatReport
}

// NewCollector creates a heartbeat telemetry collector.
func NewCollector(tenantID, licenseID string, intervalSeconds int, callback ReportCallback) *Collector {
	return &Collector{
		tenantID:      tenantID,
		licenseID:     licenseID,
		interval:      time.Duration(intervalSeconds) * time.Second,
		callback:      callback,
		events:        make([]PredictionEvent, 0, 1000),
		intervalStart: time.Now(),
		stopCh:        make(chan struct{}),
	}
}

// Record adds a prediction event. Thread-safe.
func (c *Collector) Record(event PredictionEvent) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.events = append(c.events, event)
}

// RecordPrediction is a convenience method to record a prediction.
func (c *Collector) RecordPrediction(modelID string, latencyMs float64, success bool) {
	c.Record(PredictionEvent{
		TenantID:  c.tenantID,
		ModelID:   modelID,
		LicenseID: c.licenseID,
		LatencyMs: latencyMs,
		Success:   success,
		Timestamp: time.Now(),
	})
}

// Flush aggregates current events into a report and resets.
func (c *Collector) Flush() *HeartbeatReport {
	c.mu.Lock()
	events := c.events
	c.events = make([]PredictionEvent, 0, 1000)
	intervalStart := c.intervalStart
	c.intervalStart = time.Now()
	c.mu.Unlock()

	if len(events) == 0 {
		return nil
	}

	c.reportCounter++
	now := time.Now()

	var successes, failures int64
	byModel := make(map[string]int64)
	latencies := make([]float64, 0, len(events))

	for _, e := range events {
		if e.Success {
			successes++
		} else {
			failures++
		}
		byModel[e.ModelID]++
		latencies = append(latencies, e.LatencyMs)
	}

	sort.Float64s(latencies)

	report := &HeartbeatReport{
		TenantID:              c.tenantID,
		LicenseID:             c.licenseID,
		ReportID:              fmt.Sprintf("%s-%d", c.tenantID, c.reportCounter),
		IntervalStart:         intervalStart,
		IntervalEnd:           now,
		TotalPredictions:      int64(len(events)),
		SuccessfulPredictions: successes,
		FailedPredictions:     failures,
		PredictionsByModel:    byModel,
		LatencyP50Ms:          percentile(latencies, 50),
		LatencyP95Ms:          percentile(latencies, 95),
		LatencyP99Ms:          percentile(latencies, 99),
	}

	c.mu.Lock()
	c.reports = append(c.reports, report)
	c.mu.Unlock()

	if c.callback != nil {
		c.callback(report)
	}

	return report
}

// Start begins periodic background reporting.
func (c *Collector) Start() {
	c.running = true
	go func() {
		ticker := time.NewTicker(c.interval)
		defer ticker.Stop()
		for {
			select {
			case <-ticker.C:
				c.Flush()
			case <-c.stopCh:
				return
			}
		}
	}()
}

// Stop halts periodic reporting and flushes remaining events.
func (c *Collector) Stop() *HeartbeatReport {
	if c.running {
		c.running = false
		close(c.stopCh)
	}
	return c.Flush()
}

// GetReports returns all collected reports.
func (c *Collector) GetReports() []*HeartbeatReport {
	c.mu.Lock()
	defer c.mu.Unlock()
	result := make([]*HeartbeatReport, len(c.reports))
	copy(result, c.reports)
	return result
}

func percentile(sorted []float64, pct int) float64 {
	if len(sorted) == 0 {
		return 0
	}
	idx := int(math.Max(0, float64(len(sorted)*pct/100)-1))
	if idx >= len(sorted) {
		idx = len(sorted) - 1
	}
	return sorted[idx]
}

File: services/licensing/heartbeat.py
"""
Heartbeat Telemetry Module for FHE-GBDT Serving

Collects counter-only usage telemetry from the on-prem runtime and
periodically reports it to the vendor's cloud control plane.

Telemetry includes ONLY:
- Prediction counts per model
- Latency percentiles
- Error counts
- License ID and tenant ID

Telemetry NEVER includes:
- Ciphertext payloads
- Feature values (encrypted or plaintext)
- Model weights or parameters
- Evaluation keys
"""

import json
import time
import threading
import logging
from dataclasses import dataclass, field, asdict
from typing import Dict, List, Optional, Callable
from collections import defaultdict

logger = logging.getLogger(__name__)

# Fields that must NEVER appear in telemetry
FORBIDDEN_FIELDS = frozenset({
    "ciphertext", "plaintext", "secret_key", "eval_key",
    "features", "predictions", "payload", "api_key",
    "password", "token", "model_weights",
})


@dataclass
class PredictionEvent:
    """A single prediction event for telemetry aggregation."""
    tenant_id: str
    model_id: str
    license_id: str
    latency_ms: float
    success: bool
    timestamp: float = field(default_factory=time.time)


@dataclass
class HeartbeatReport:
    """Aggregated telemetry report sent to the control plane."""
    tenant_id: str
    license_id: str
    report_id: str
    interval_start: float
    interval_end: float
    total_predictions: int
    successful_predictions: int
    failed_predictions: int
    predictions_by_model: Dict[str, int]
    latency_p50_ms: float
    latency_p95_ms: float
    latency_p99_ms: float

    def to_dict(self) -> Dict:
        d = asdict(self)
        # Safety: ensure no forbidden fields leak
        for key in list(d.keys()):
            if key in FORBIDDEN_FIELDS:
                del d[key]
        return d

    def to_json(self) -> str:
        return json.dumps(self.to_dict())


class HeartbeatCollector:
    """
    Collects prediction events and produces periodic heartbeat reports.

    Runs on the customer's on-prem infrastructure. Reports are counters only --
    no sensitive data is ever included.
    """

    def __init__(
        self,
        tenant_id: str,
        license_id: str,
        report_interval_seconds: int = 60,
        report_callback: Optional[Callable[[HeartbeatReport], None]] = None,
    ):
        self._tenant_id = tenant_id
        self._license_id = license_id
        self._interval = report_interval_seconds
        self._callback = report_callback

        self._lock = threading.Lock()
        self._events: List[PredictionEvent] = []
        self._interval_start = time.time()
        self._report_counter = 0

        self._running = False
        self._timer: Optional[threading.Timer] = None
        self._reports: List[HeartbeatReport] = []

    def record(self, event: PredictionEvent) -> None:
        """Record a prediction event. Thread-safe."""
        with self._lock:
            self._events.append(event)

    def record_prediction(
        self,
        model_id: str,
        latency_ms: float,
        success: bool = True,
    ) -> None:
        """Convenience method to record a prediction."""
        self.record(PredictionEvent(
            tenant_id=self._tenant_id,
            model_id=model_id,
            license_id=self._license_id,
            latency_ms=latency_ms,
            success=success,
        ))

    def flush(self) -> Optional[HeartbeatReport]:
        """Flush current events and produce a report."""
        with self._lock:
            if not self._events:
                return None

            events = self._events
            self._events = []
            interval_start = self._interval_start
            self._interval_start = time.time()

        now = time.time()
        self._report_counter += 1

        # Aggregate
        total = len(events)
        successes = sum(1 for e in events if e.success)
        failures = total - successes

        by_model: Dict[str, int] = defaultdict(int)
        for e in events:
            by_model[e.model_id] += 1

        latencies = sorted(e.latency_ms for e in events)
        p50 = self._percentile(latencies, 50)
        p95 = self._percentile(latencies, 95)
        p99 = self._percentile(latencies, 99)

        report = HeartbeatReport(
            tenant_id=self._tenant_id,
            license_id=self._license_id,
            report_id=f"{self._tenant_id}-{self._report_counter}",
            interval_start=interval_start,
            interval_end=now,
            total_predictions=total,
            successful_predictions=successes,
            failed_predictions=failures,
            predictions_by_model=dict(by_model),
            latency_p50_ms=p50,
            latency_p95_ms=p95,
            latency_p99_ms=p99,
        )

        self._reports.append(report)

        if self._callback:
            try:
                self._callback(report)
            except Exception as exc:
                logger.warning("Heartbeat callback failed: %s", exc)

        return report

    def start(self) -> None:
        """Start periodic reporting in a background thread."""
        self._running = True
        self._schedule_next()

    def stop(self) -> Optional[HeartbeatReport]:
        """Stop periodic reporting and flush remaining events."""
        self._running = False
        if self._timer:
            self._timer.cancel()
            self._timer = None
        return self.flush()

    def get_reports(self) -> List[HeartbeatReport]:
        """Return all collected reports."""
        return list(self._reports)

    def _schedule_next(self) -> None:
        if not self._running:
            return
        self._timer = threading.Timer(self._interval, self._tick)
        self._timer.daemon = True
        self._timer.start()

    def _tick(self) -> None:
        self.flush()
        self._schedule_next()

    @staticmethod
    def _percentile(sorted_values: List[float], pct: int) -> float:
        if not sorted_values:
            return 0.0
        idx = max(0, int(len(sorted_values) * pct / 100) - 1)
        return sorted_values[idx]
